### PR TITLE
fix 4425: resolve --basetemp to absolute paths

### LINKED ISSUE #4425: resolve manually set basetemp to a absolute path
Hi Guys,

We're experiencing some issues with new tmpdir implementation using pathlib. Starting from pytest 3.9.0 our tests started to fail. 

We are using test class and a pre_process fixture, which prepares some data for the other tests within the class. In the test methods we're using the pre_process fixture (to access the data) as well as the tmpdir pytest builtin fixture to store some outcomes of the tests. Minimal failing example is pasted below:
```
import pytest, os

class TestDummy():
    @pytest.fixture(scope="class")
    def pre_process(self, tmpdir_factory):
        path_with_data = tmpdir_factory.mktemp('data')
        return path_with_data

    @pytest.mark.parametrize("param", range(2))
    def test_dummy(self, tmpdir, pre_process, param):
        os.chdir(pre_process)
        # further processing

    def test_dummy2(self, tmpdir, pre_process):
        os.chdir(pre_process)
        # further processing

```
Because of the directory change in test_dummy, next invocation of the tests fail due to pathlib error, as below:
```____________________________________________________________________________________________________ ERROR at setup of TestDummy.test_dummy[1] _____________________________________________________________________________________________________

self = PosixPath('tmp')

    def iterdir(self):
        """Iterate over the files in this directory.  Does not yield any
        result for the special paths '.' and '..'.
        """
        if self._closed:
            self._raise_closed()
>       for name in self._accessor.listdir(self):

../../../.pyenv/versions/3.6.5/lib/python3.6/pathlib.py:1079: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

pathobj = PosixPath('tmp'), args = ()

    @functools.wraps(strfunc)
    def wrapped(pathobj, *args):
>       return strfunc(str(pathobj), *args)
E       FileNotFoundError: [Errno 2] No such file or directory: 'tmp'

../../../.pyenv/versions/3.6.5/lib/python3.6/pathlib.py:387: FileNotFoundError
```

and in fact we end up with 1 pass and 2 errors in this particular case.

We currently have workaround for this issue (store the cwd before changing it and restore it before the test end), but I'd like to ask you if this is how it should behave? I'd expect that state of one test does not influence the other test's state. 

The issue is observable if no class hierarchy is used as well. 

Another workaround (a little bit nasty) is to reset internal pytest variable in pre_process macro:
```@pytest.fixture(scope="module")
def pre_process(tmpdir_factory):
    path_with_data = tmpdir_factory.mktemp('data')
    tmpdir_factory._tmppath_factory._basetemp = None
    return path_with_data```

Environment details:
Python 3.6.5 (via pyenv), 4.15.0-29-generic #31-Ubuntu SMP Tue Jul 17 15:37:27 UTC 2018 i686 i686 i686 GNU/Linux
```
$ pip list
Package        Version
-------------- -------
atomicwrites   1.2.1  
attrs          18.2.0 
more-itertools 4.3.0  
pip            18.1   
pluggy         0.8.0  
py             1.7.0  
pytest         4.0.0  
setuptools     40.6.2 
six            1.11.0 ```

- [X] Include a detailed description of the bug or suggestion
- [X] `pip list` of the virtual environment you are using:
- [X] pytest and operating system versions:
- [X] Minimal example if possible - already in the text

### PR DESCRIPTION
**Pull request by RonnyPfannschmidt**
fixes #4425

### CODE REVIEW

File: testing/test_tmpdir.py [hunk 1]
```diff
@@ -25,12 +26,29 @@ def test_ensuretemp(recwarn):
     assert d1.check(dir=1)
 
 
+@attr.s
+class FakeConfig(object):
```
Comment (nicoddemus) [inline testing/test_tmpdir.py]: TBH not sure if this fake actually improves things... the "option returns self" and `get` are quite the hack. 😬

File: testing/test_tmpdir.py [hunk 2]
```diff
@@ -40,6 +58,14 @@ def test_mktemp(self, testdir):
         assert tmp2.relto(t.getbasetemp()).startswith("this")
         assert tmp2 != tmp
 
+    def test_tmppath_relative_basetemp_absolute(self, tmp_path, monkeypatch):
+        from _pytest.tmpdir import TempPathFactory
```
Comment (nicoddemus) [inline testing/test_tmpdir.py]: I would add a docstring and a reference to the bug. 👍

File: testing/test_tmpdir.py (hunk 1 above)
Comment (RonnyPfannschmidt) [reply testing/test_tmpdir.py]: its indeed quite a hack, we cant make a sane config object for use in such cases

**Comment by nicoddemus** [comment]
Found a discussion [about `Path.absolute()`](https://bugs.python.org/issue25012), it is really public by accident.

Fix the `os.path.abspath` call to handle the case where the parameter is already a path, in Path-clueless versions.

**Comment by RonnyPfannschmidt** [comment]
restarted the build, travis was broken

**Comment by nicoddemus** [comment]
Weird, on MacOS the test fails, here's the exception formatted for readability:

```python
PosixPath('/private/var/folders/nz/vv4_9tw56nv9k3tkvyszvwg80000gn/T/pytest-of-travis/pytest-0/test_tmppath_relative_basetemp0/hello')
 == 
PosixPath('/var/folders/nz/vv4_9tw56nv9k3tkvyszvwg80000gn/T/pytest-of-travis/pytest-0/test_tmppath_relative_basetemp0') / 'hello'
```

Seems to be symlink-related...

**Comment by RonnyPfannschmidt** [comment]
it is symlink related, other projects see the same issue

we can use resolve on the resulting paths

**Comment by nicoddemus** [comment]
ahh good catch!

**Comment by pllim** [comment]
Hello. I am looking forward to this fix. Which release is it going to be in? Thanks.

**Comment by RonnyPfannschmidt** [comment]
the next release in any case

File: src/_pytest/tmpdir.py [hunk 3]
```diff
@@ -26,7 +27,14 @@ class TempPathFactory(object):
 
     The base directory can be configured using the ``--basetemp`` option."""
 
-    _given_basetemp = attr.ib()
+    _given_basetemp = attr.ib(
+        # using os.path.abspath() to get absolute path instead of resolve() as it
+        # does not work the same in all platforms (see #4427)
+        # Path.absolute() exists, but it is not public (see https://bugs.python.org/issue25012)
+        convert=attr.converters.optional(
+            lambda p: Path(os.path.abspath(six.text_type(p)))
```
Comment (techalchemy) [inline src/_pytest/tmpdir.py]: Brief note (was stalking IRC so I saw this comment) -- `resolve()` doesn't only resolve symlinks, but on windows it also resolves UNC paths from mounts (e.g. if you mount a network share at `R:\\Some\\path` but it really points at `\\10.1.100.100\Some\path`, `Path.resolve()` captures that but `os.path.abspath()` will return `R:\\blah` -- despite `absolute` not being public we typically handle this by trying `resolve` and falling back on `absolute` in pipenv at least

Comment (nicoddemus) [reply src/_pytest/tmpdir.py]: Thanks @techalchemy for the note, that's interesting! 👍

**Pull request merged**